File: src/dataset/cpp/run-length-encoding_exercism-chrisnp/gemini/GeminiFlash_run_length_encoding_v3.cpp
#include "run_length_encoding.h"

#include <algorithm>
#include <cctype>
#include <string>
#include <vector>

using std::string;
using std::to_string;

string run_length_encoding::encode(const string &decoded) noexcept {
    if (decoded.empty()) return {};

    string encoded;
    encoded.reserve(decoded.length()); 

    char current_char = decoded[0];
    int count = 1;

    for (size_t i = 1; i < decoded.length(); ++i) {
        if (decoded[i] == current_char) {
            ++count;
        } else {
            if (count > 1) {
                encoded += to_string(count);
            }
            encoded.push_back(current_char);
            current_char = decoded[i];
            count = 1;
        }
    }

    if (count > 1) {
        encoded += to_string(count);
    }
    encoded.push_back(current_char);

    return encoded;
}

string run_length_encoding::decode(const string &encoded) noexcept {
    if (encoded.empty()) return {};

    string decoded;
    decoded.reserve(encoded.length());

    int count = 0;
    for (char ch : encoded) {
        if (isdigit(ch)) {
            count = count * 10 + (ch - '0');
        } else {
            if (count == 0) {
                decoded.push_back(ch);
            } else {
                decoded.append(count, ch);
                count = 0;
            }
        }
    }
    return decoded;
}

File: src/dataset/javascript/run-length-encoding_exercism-ThomasZumsteg/gemini/GeminiFlash_run-length-encoding_v3.js
module.exports = {
    encode: function(text) {
        let encoded = "";
        let count = 1;
        for (let i = 0; i < text.length; i++) {
            if (i + 1 < text.length && text[i] === text[i + 1]) {
                count++;
            } else {
                encoded += (count > 1 ? count : "") + text[i];
                count = 1;
            }
        }
        return encoded;
    },
    decode: function(text) {
        let decoded = "";
        let i = 0;
        while (i < text.length) {
            let numStr = "";
            while (i < text.length && /\d/.test(text[i])) {
                numStr += text[i];
                i++;
            }
            const char = text[i];
            const num = numStr ? parseInt(numStr, 10) : 1;
            decoded += char.repeat(num);
            i++;
        }
        return decoded;
    }
};